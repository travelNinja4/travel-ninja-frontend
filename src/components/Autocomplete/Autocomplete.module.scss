/**
------------------------------------------------------
styles specific to the Autocomplete component
------------------------------------------------------
**/

.dropdown {
  position: relative;
  width: 16rem;
}

.inputWrapper {
  position: relative;
  display: flex;
  align-items: center;
  width: 100%;
}

.input {
  width: 100%;
  padding: 0.75rem 0.87rem;
  border: 1px solid var(--color-gray-300);
  border-radius: 6px;
  background: var(--color-white);
  font-size: 1rem;
  background-color: var(--color-white);
  color: var(--color-black);
  color-scheme: light;
  transition:
    border-color 0.2s ease,
    box-shadow 0.2s ease;

  &:focus {
    outline: none;
    border-color: var(--color-indigo);
    box-shadow: 0 0 0 0.2rem var(--color-indigo-alpha-25);
  }

  &:disabled {
    background: var(--color-gray-120);
    cursor: not-allowed;
    color: var(--color-gray-350);
  }
}

.list {
  position: absolute;
  top: calc(100% + 0.25rem);
  left: 0;
  width: 100%;
  margin: 0;
  padding: 0;
  list-style: none;
  border: 1px solid var(--color-gray-220);
  border-radius: 6px;
  background: var(--color-white);
  box-shadow: 0 4px 8px var(--color-black-alpha-10);
  max-height: 12rem;
  overflow-y: auto;
  z-index: 100;

  // smooth open/close animation
  animation: fadeIn 0.15s ease-in-out;
}

.option {
  padding: 0.5rem 0.75rem;
  cursor: pointer;
  font-size: 0.95rem;
  color: var(--color-gray-700);
  transition:
    background 0.2s ease,
    color 0.2s ease;

  &:hover,
  &.highlighted {
    background: var(--color-option-hover-bg);
  }

  &.selected {
    font-weight: 600;
    background: var(--color-option-selected-bg);
    color: var(--color-option-selected-text);
  }
}

.noOptions {
  padding: 0.5rem 0.75rem;
  font-size: 0.9rem;
  color: var(--color-gray-350);
  text-align: center;
}

.loading {
  display: flex;
  align-items: center;
  gap: 0.5rem;
  color: var(--color-gray-320);
  font-size: 0.875rem;
  height: 45px;
  .spinner {
    height: 1rem;
    width: 1rem;
    border: 2px solid var(--icon-color-info);
    border-top-color: transparent;
    border-radius: 9999px;
    animation: spin 1s linear infinite;
  }
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

@keyframes fadeIn {
  from {
    opacity: 0;
    transform: translateY(-4px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}
