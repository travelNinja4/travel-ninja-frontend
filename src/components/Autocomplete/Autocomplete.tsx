/**
 * A component that allows users to search, filter, and select options from a list with dynamic suggestions as they type.
 *
 * @example
 * ```tsx
 * import Autocomplete from '@src/components/Autocomplete'
 *
 * export default function Autocomplete() {
 *   return <Autocomplete label="Hello" />;
 * }
 * ```
 */
'use client';

import { useState, useRef, useEffect, KeyboardEvent, ChangeEvent } from 'react';
import clsx from 'clsx';
import styles from './Autocomplete.module.scss';

export type AutocompleteOption<T> = {
  value: T;
  label: string;
};

/**
 * Define the props available for the Autocomplete component.
 */
interface AutocompleteProps<T> {
  options: AutocompleteOption<T>[];
  value?: T | string;
  onChange?: (value: T | string | undefined) => void;
  searchValue?: string;
  onSearchChange?: (value: string) => void;
  placeholder?: string;
  dropdownContainerClassName?: string;
  dropdownTextContainerClassName?: string;
  disabled?: boolean;
}

export default function Autocomplete<T>({
  options,
  value,
  onChange,
  searchValue = '',
  onSearchChange,
  placeholder = 'Search...',
  dropdownContainerClassName,
  dropdownTextContainerClassName,
  disabled = false,
}: AutocompleteProps<T>) {
  const [highlightedIndex, setHighlightedIndex] = useState<number>(0);
  const [isOpen, setIsOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);

  // Close dropdown when clicking outside
  useEffect(() => {
    const handleClickOutside = (e: MouseEvent) => {
      if (dropdownRef.current && !dropdownRef.current.contains(e.target as Node)) {
        setIsOpen(false);
      }
    };
    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, []);

  // Keyboard navigation
  const handleKeyDown = (e: KeyboardEvent<HTMLInputElement | HTMLUListElement>) => {
    if (!isOpen) return;

    if (e.key === 'ArrowDown') {
      setHighlightedIndex((prev) => (prev + 1) % options.length);
      e.preventDefault();
    } else if (e.key === 'ArrowUp') {
      setHighlightedIndex((prev) => (prev - 1 + options.length) % options.length);
      e.preventDefault();
    } else if (e.key === 'Enter') {
      const option = options[highlightedIndex];
      onChange?.(option.value);
      setIsOpen(false);
      e.preventDefault();
    } else if (e.key === 'Escape') {
      setIsOpen(false);
      e.preventDefault();
    }
  };

  return (
    <div ref={dropdownRef} className={clsx(styles.dropdown, dropdownContainerClassName)}>
      <div className={styles.inputWrapper}>
        <input
          type="text"
          placeholder={placeholder}
          value={searchValue || (value as string)}
          className={clsx(styles.input, dropdownTextContainerClassName)}
          onFocus={() => !disabled && setIsOpen(true)}
          onChange={(e: ChangeEvent<HTMLInputElement>) => {
            if (disabled) return;
            onSearchChange?.(e.target.value);
            setIsOpen(true);
          }}
          onKeyDown={handleKeyDown}
          disabled={disabled}
        />
      </div>

      {isOpen && options.length > 0 && (
        <ul role="listbox" tabIndex={-1} onKeyDown={handleKeyDown} className={styles.list}>
          {options.map((option, idx) => {
            const isHighlighted = idx === highlightedIndex;
            return (
              <li
                key={idx}
                role="option"
                className={clsx(styles.option, { [styles.highlighted]: isHighlighted })}
                onMouseEnter={() => setHighlightedIndex(idx)}
                onClick={() => {
                  onChange?.(option.value);
                  setIsOpen(false);
                }}
              >
                {option.label}
              </li>
            );
          })}
        </ul>
      )}

      {isOpen && options.length === 0 && (
        <ul className={styles.list}>
          <li className={styles.noOptions}>No options</li>
        </ul>
      )}
    </div>
  );
}
