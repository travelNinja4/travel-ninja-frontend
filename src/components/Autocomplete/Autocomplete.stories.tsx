import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import Autocomplete from './Autocomplete'

const meta = {
  title: 'components/Autocomplete',
  component: Autocomplete,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    label: { 
        control: 'text', 
        description: 'Sample label for the component' 
        },
  },
  args: { 
    /** onClick: fn() **/
    },
} satisfies Meta<typeof Autocomplete>;

export default meta;
type Story = StoryObj<typeof meta>;