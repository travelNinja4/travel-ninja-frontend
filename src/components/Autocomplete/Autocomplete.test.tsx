import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import Autocomplete from './Autocomplete';

describe('Home', () => {
  /** Base props for Autocomplete **/
  const baseProps = {
    options: [
      { value: 'apple', label: 'Apple' },
      { value: 'banana', label: 'Banana' },
      { value: 'cherry', label: 'Cherry' },
    ],
    value: '',
    onChange: jest.fn(),
    searchValue: '',
    onSearchChange: jest.fn(),
    placeholder: 'Search...',
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  /**
   * @test
   * Renders the Autocomplete component without throwing an error
   * and checks if the AutocompleteTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<Autocomplete {...baseProps} />);
    expect(screen.getByTestId('AutocompleteTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the Autocomplete component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<Autocomplete {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the Autocomplete component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<Autocomplete {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('opens dropdown on focus and shows options', () => {
    render(<Autocomplete {...baseProps} />);
    const input = screen.getByPlaceholderText('Search...');
    fireEvent.focus(input);
    expect(screen.getByRole('listbox')).toBeInTheDocument();
    expect(screen.getAllByRole('option')).toHaveLength(3);
  });

  it('calls onSearchChange when typing', () => {
    render(<Autocomplete {...baseProps} />);
    const input = screen.getByPlaceholderText('Search...');
    fireEvent.change(input, { target: { value: 'ap' } });
    expect(baseProps.onSearchChange).toHaveBeenCalledWith('ap');
  });

  it('selects an option on click', () => {
    render(<Autocomplete {...baseProps} />);
    const input = screen.getByPlaceholderText('Search...');
    fireEvent.focus(input);
    const option = screen.getByText('Banana');
    fireEvent.click(option);
    expect(baseProps.onChange).toHaveBeenCalledWith('banana');
  });

  it('navigates options with keyboard and selects with Enter', () => {
    render(<Autocomplete {...baseProps} />);
    const input = screen.getByPlaceholderText('Search...');
    fireEvent.focus(input);
    fireEvent.keyDown(input, { key: 'ArrowDown' });
    fireEvent.keyDown(input, { key: 'Enter' });
    expect(baseProps.onChange).toHaveBeenCalledWith('banana');
  });

  it('closes dropdown on Escape key', () => {
    render(<Autocomplete {...baseProps} />);
    const input = screen.getByPlaceholderText('Search...');
    fireEvent.focus(input);
    expect(screen.getByRole('listbox')).toBeInTheDocument();
    fireEvent.keyDown(input, { key: 'Escape' });
    expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
  });

  it('does not open dropdown when disabled', () => {
    render(<Autocomplete {...baseProps} disabled />);
    const input = screen.getByPlaceholderText('Search...');
    fireEvent.focus(input);
    expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
  });

  it('closes when clicking outside', () => {
    render(
      <div>
        <Autocomplete {...baseProps} />
        <button data-testid="outside">Outside</button>
      </div>,
    );
    const input = screen.getByPlaceholderText('Search...');
    fireEvent.focus(input);
    expect(screen.getByRole('listbox')).toBeInTheDocument();
    fireEvent.mouseDown(screen.getByTestId('outside'));
    expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
  });

  it('navigates options with ArrowUp and selects with Enter', () => {
    render(<Autocomplete {...baseProps} />);
    const input = screen.getByPlaceholderText('Search...');
    fireEvent.focus(input);

    fireEvent.keyDown(input, { key: 'ArrowUp' });
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(baseProps.onChange).toHaveBeenCalledWith('cherry');
  });
});
