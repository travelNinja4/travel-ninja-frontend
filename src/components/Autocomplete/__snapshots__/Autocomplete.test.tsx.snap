// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <div
      class="inputWrapper"
    >
      <input
        class="input"
        placeholder="Search..."
        type="text"
      />
    </div>
  </div>
</DocumentFragment>
`;
